### ISSUE TITLE
OpenSSL crypto driver IO failure when large IO

### ISSUE THREAD

**Issue by yin19941005** [spdk/spdk#2759]
Hello,

# Sighting report

The issue only affects OpenSSL crypto driver. The bdev will be set to SPDK_BDEV_IO_STATUS_FAILED and the IO operation will fail when the re-enqueue is started. When a large amount of IO occurs, the driver is no longer able to enqueue all of them at once. The driver starts trying to re-enqueue (re-submit) the operation after dequeue and OpenSSL crypto driver will fail.

## Expected Behavior

The OpenSSL crypto driver should remain functional even if it starts to re-enqueue, just as the AESNI_MB and MLX5 crypto drivers.

## Current Behavior

The OpenSSL crypto driver will fail the IO operation when it starts to re-enqueue.

## Possible Solution

1. Handling the status `RTE_CRYPTO_OP_STATUS_SUCCESS` in re-enqueue as well. And do not re-perform the crypto operation in the crypto driver.

Please check my commit on fork and the PR:

SPDK part:
https://github.com/yin19941005/spdk/commit/e31b388ba6d5311733097d4078e4a0f2345f1117

DPDK part:
https://github.com/yin19941005/spdk-dpdk/commit/db3b329fa6e9d0cf0b40a45d396c111bad16c291

PR: #2760 

## Steps to Reproduce

Prerequisite: a fast environment for reproduce. From my experience, if using NVMe-oF with 1Gb ethernet, almost impossible to reproduce the issue. I am using 100Gb ethernet and spotted the issue.

For OpenSSL crypto driver,
1. Enable the OpenSSL crypto driver in the configure file. For me, I added the OpenSSL crypto driver on crypto option.
2. Make the SPDK with OpenSSL crypto driver. (As I added the OpenSSL crypto driver in crypto option, I simply add the crypto option)
3. Start the SPDK and start OpenSSL crypto driver
4. Monitoring and launching a large amount of IO

For reference, to setup AESNI_MB crypto driver

1. Make the SPDK with crypto option
2. Start the SPDK and start AESNI_MB crypto driver
3. Monitoring and launching a large amount of IO

## Context (Environment including OS version, SPDK version, etc.)

I am using NVMe-oF environment on 100Gb network for the testing and multiple testing platform are tested, 
tested SPDK version: v22.05, v22.09, v23.01-pre

NVMe-oF initiator:
Using NIC: NVIDIA BLUEFIELD-2 DPU
OS version= "Ubuntu 20.04.3 LTS (Focal Fossa)" (not sure NVIDIA BLUEFIELD-2 DPU got any customization on it or not)

Another NVMe-oF initiator
Using NIC: NVIDIA Mellanox ConnectX-5
OS version= Rocky Linux 8.6

For reference, NVMe-oF target:
Using NIC: NVIDIA Mellanox ConnectX-5
OS version= Rocky Linux 8.6
tested SPDK version: v22.05, v22.09, v23.01-pre

## Comment

Hi @AlekseyMarchuk, this is the issue I mentioned on #2756 yesterday. If you are working on refactoring the crypto-related part of SPDK, could you also address this issue as well?

The current SPDK `vbdev_crypto.c` is only checking `RTE_CRYPTO_OP_STATUS_NOT_PROCESSED` on re-enqueue.

https://github.com/spdk/spdk/blob/master/module/bdev/crypto/vbdev_crypto.c#L745

However, in the OpenSSL crypto driver implementation, it is going to set the operation status to `RTE_CRYPTO_OP_STATUS_SUCCESS` before actually enqueuing it.

https://github.com/spdk/dpdk/blob/db7634dee3b8eac6817367454864ee6bd071263e/drivers/crypto/openssl/rte_openssl_pmd.c#L2788

As a result, when the re-enqueue occurs, the OpenSSL crypto driver is performing the crypto operation again and causing **data corruption**. This is because OpenSSL crypto driver didn't check whether the operation is already done or on re-enqueue.

**Comment by Comphix** [comment]
Hi @yin19941005 , thanks for raising the issue and also the potential change.

Could you follow the https://spdk.io/development/ website for the PR (patch) of review?

SPDK's ongoing changes are on Gerrit: https://review.spdk.io/gerrit/dashboard/self

**Comment by AlekseyMarchuk** [comment]
Hi @yin19941005 
Thank you for reporting this issue. There is another bug #2907 which is related to OpenSSL and RTE_CRYPTO_OP_STATUS_SUCCESS crypto op status. I pushed a fix which needs to be reworked a bit. I'll update it tomorrow and will notify you.
Since this is a duplicate of #2907, I'm closing this issue

**Comment by yin19941005** [comment]
Hi @AlekseyMarchuk,

Thank you for helping! I will keep track on #2907 for this issue. Thank you!
